### ISSUE TITLE
Pyright doesn't seem to handle staticmethod and classmethod decorators correctly

### ISSUE THREAD

**Issue by QuantumTim** [microsoft/pyright#9108]
**Describe the bug**
Pyright doesn't seem to handle `staticmethod` and `classmethod` decorators correctly. If you apply the `@staticmethod` decorator to a function, then it types that name as a plain callable, rather than a `staticmethod` instance. However if you manually call `staticmethod(fn)` then you get a correctly typed object.

**Code or Screenshots**
Minimal example:
```python
from typing import reveal_type

class Foo:
    @staticmethod
    def f(a: int, *, b: str) -> dict[bool, bool]:
        return {False: True}

    def _g(a: int, *, b: str) -> dict[bool, bool]: # <-- pyright complains about missing self parameter
        return {True: False}
    g = staticmethod(_g)

    reveal_type(f)     # Runtime: staticmethod; pyright: (a: int, *, b: str) -> dict[bool, bool]
    reveal_type(g)     # Runtime: staticmethod; pyright: staticmethod[(a: int, *, b: str), dict[bool, bool]]
```

What I'm actually trying to do is create my own `method` function decorator. The example below is a bit dumb, in that it doesn't really do anything more than normal Python functions/methods, but the real code adds some extra functionality when calling such a method. The typing should behave like normal methods though (ie applying the `@method` decorator should type the same as a normal method).

For normal methods, this works fine, but I'd also like to support the equivalent of `staticmethod` and `classmethod`. I've tried various ways of doing this, but I can't get anything to work with pyright. The simplest technique I can think of is just to combine my `@method` decorator with the built-in `@staticmethod` and `@classmethod` decorators, which is what I've shown below. But because of the way pyright currently handles these, it then gets all the typing wrong on my stuff.

An alternative implementations I tried was to just create my own equivalent `@my_staticmethod` decorator. I can get this to type correctly from the outside, but pyright will then complain about all my functions missing the "self" parameter, because I have no way to tell pyright to treat the inner function definition as a staticmethod (or similarly for classmethods).

Apologies for the long snippet here. The short snippet above while technically demonstrating the problem, doesn't actually show any consequences of it, and this was about as small as I could get to make it work for real, but still demonstrate the issues.

```python
from __future__ import annotations
from typing import Any, Callable, Concatenate, Generic, ParamSpec, Self, TypeVar, overload, reveal_type

_P = ParamSpec('_P')
_Q = ParamSpec('_Q')
_S_contra = TypeVar('_S_contra', contravariant=True)
_T_co = TypeVar('_T_co', covariant=True)
_U_co = TypeVar('_U_co', covariant=True)

class bound_method(Generic[_S_contra, _P, _T_co]):
    def __new__(cls, instance: _S_contra, m: method[Concatenate[_S_contra, _P], _T_co]) -> Self:
        self = super().__new__(cls)
        self._instance = instance
        self._method = m
        return self
    def __call__(self, /, *args: _P.args, **kwargs: _P.kwargs) -> _T_co:
        self._instance: _S_contra                               # Work around for issue #9107
        self._method: method[Concatenate[_S_contra, _P], _T_co] # Work around for issue #9107
        return self._method.__call__(self._instance, *args, **kwargs)

class method(Generic[_P, _T_co]):
    _fn: Callable[_P, _T_co]
    def __new__(cls, fn: Callable[_P, _T_co]) -> Self:
        self = super().__new__(cls)
        self._fn = fn
        return self
    @overload
    def __get__(self, instance: None, owner: type[Any]) -> Self: ...
    @overload
    def __get__(self: method[Concatenate[_S_contra, _Q], _U_co], instance: _S_contra, owner: type[_S_contra]|None = None) -> bound_method[_S_contra, _Q, _U_co]: ...
    def __get__(self, instance: Any|None, owner: type[Any]|None = None) -> bound_method|Self:
        if instance is None:
            return self
        return bound_method(instance, self) # type: ignore
    def __call__(self, /, *args: _P.args, **kwargs: _P.kwargs) -> _T_co:
        return self._fn(*args, **kwargs)

class Foo:
    @method
    def meth(self, a: int, *, b: str) -> list[bool]:
        return [True]

    @staticmethod
    @method
    def smeth(a: int, *, b: str) -> tuple[bool]:
        return (True,)
    #smeth = staticmethod(_smeth) <-- Non-decorator syntax works, but is messier/uglier

    @classmethod
    @method
    def cmeth(cls, a: int, *, b: str) -> set[bool]:
        return {True}
    #cmeth = classmethod(_cmeth) <-- Non-decorator syntax works, but is messier/uglier

    reveal_type(meth)  # Runtime: method;       pyright: method[(self: Self@Foo, a: int, *, b: str), list[bool]]
    reveal_type(smeth) # Runtime: staticmethod; pyright: method[(a: int, *, b: str), tuple[bool]]
    reveal_type(cmeth) # Runtime: classmethod;  pyright: method[(cls: type[Self@Foo], a: int, *, b: str), set[bool]]

foo = Foo()

print('\nMethod:')
reveal_type(Foo.meth)                  # Runtime: method;       pyright: method[(self: Foo, a: int, *, b: str), list[bool]]
reveal_type(foo.meth)                  # Runtime: bound_method; pyright: bound_method[Foo, (a: int, *, b: str), list[bool]]
reveal_type(Foo.meth(foo, 1, b='one')) # Runtime: list;         pyright: list[bool]
reveal_type(foo.meth(1, b='one'))      # Runtime: list;         pyright: list[bool]
print('\nStatic Method:')
reveal_type(Foo.smeth)                 # Runtime: method;       pyright: method[(a: int, *, b: str), tuple[bool]]
reveal_type(foo.smeth)                 # Runtime: method;       pyright: Unknown (error: cannot access attribute "smeth")
reveal_type(Foo.smeth(1, b='one'))     # Runtime: tuple;        pyright: tuple[bool]
reveal_type(foo.smeth(1, b='one'))     # Runtime: tuple;        pyright: Unknown (error: cannot access attribute "smeth")
print('\nClass Method:')
reveal_type(Foo.cmeth)                 # Runtime: bound_method; pyright: method[(cls: type[Foo], a: int, *, b: str), set[bool]]
reveal_type(foo.cmeth)                 # Runtime: bound_method; pyright: Unknown (error: cannot access attribute "cmeth")
reveal_type(Foo.cmeth(1, b='one'))     # Runtime: set;          pyright: Unknown (error: Argument missing for parameter "a")
reveal_type(foo.cmeth(1, b='one'))     # Runtime: set;          pyright: Unknown (error: cannot access attribute "cmeth")

```

~Interestingly, I assumed that getting a "bound staticmethod" (respectively classmethod) would be some specific built-in type, but in practice it looks like it just calls the `__get__` of the wrapped function/object with `None` for the instance argument (respectively classmethods pass the `owner` through to the underlying descriptor as the `instance`). This means that accessing those members actually returns instances of my `method` and `bound_method` types. If you use the non-descriptor syntax to make pyright "work" in this case, it doesn't know this, and just returns plain callable types for `foo.smeth`, `foo.cmeth` etc.~ Ignore this - apparently this was a "feature" that was added in 3.8 but deprecated in 3.11 and removed in 3.13.

**VS Code extension or command-line**
Using both the pyright VSCode plugin (v1.1.382), and pyright from PyPI on the command-line (1.1.382).

**Comment by erictraut** [comment]
The `@staticmethod` and `@classmethod` decorators (along with a few others like `@dataclass` and `@property`) require significant special casing in Python static type checkers. To be understood, they must be applied as decorators on a function declared with a `def` statement. If you apply them elsewhere, as you're doing in your code sample above, the special-casing cannot be applied, and you end up with an object that is simply a `staticmethod` or `classmethod` object. 

To support your use case, I think the type system would need to be extended in some way. There is currently no way in the type system to represent the difference between a callable that is a static method, a class method, or an instance method. These different types of callables are tracked internally by type checkers, but there's no way to "spell" these in type expressions. 

Not surprisingly, you'll see the same behavior with mypy and other Python static type checkers.

If you'd like to pursue some extensions to the type system to accommodate this use case, the [Python typing forum](https://discuss.python.org/c/typing/32) would be a good place to start the conversation. Ultimately, you'd need to come up with a proposal that the community agrees upon and a spec that can be incorporated into the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html).

In the meantime, I'm going to close this issue because pyright is working as intended here.

**Comment by QuantumTim** [comment]
Just for reference (in case I do decide to propose something), what sort of special casing do `@staticmethod` and `@classmethod` need? It's fairly straightforward to make your own custom `@mystaticmethod` or `@myclassmethod` (or I think even `@myproperty` for that matter - though agree that stuff like `@mydataclass` is harder - at least for typing purposes). As far as I can tell with my own implementation, the only bit that pyright doesn't like is that it doesn't know what the first argument of my function should be, and actively complains if I try to type is correctly:

```python
class A:
    @myclassmethod
    # error: Type of parameter "cls" must be a supertype of its class "A"
    def clsmeth(cls: type[Self], a: int, *, b: str) -> list[bool]:
        return [True]

    @mystaticmethod
    # error: Type of parameter "a" must be a supertype of its class "A"
    def staticmeth(a: int, *, b: str) -> set[bool]:
        return {False}

# But regardless of the complaints above, it type checks all of this correctly.
a = A()
reveal_type(A.clsmeth)              # Runtime: bound_method; pyright: bound_method[type[A], (a: int, *, b: str), list[bool]]
reveal_type(a.clsmeth)              # Runtime: bound_method; pyright: bound_method[type[A], (a: int, *, b: str), list[bool]]
reveal_type(A.clsmeth(1, b='2'))    # Runtime: list; pyright: list[bool]
reveal_type(a.clsmeth(1, b='2'))    # Runtime: list; pyright: list[bool]
reveal_type(A.staticmeth)           # Runtime: function; pyright: (a: int, *, b: str) -> set[bool]
reveal_type(a.staticmeth)           # Runtime: function; pyright: (a: int, *, b: str) -> set[bool]
reveal_type(A.staticmeth(1, b='2')) # Runtime: set; pyright: set[bool]
reveal_type(a.staticmeth(1, b='2')) # Runtime: set; pyright: set[bool]
```

Of course I wouldn't expect pyright to know what the `cls` parameter in `clsmeth` is supposed to be - though maybe in the future there could be some `InstanceSelf` magic typeform thing (or `type[InstanceSelf]` for a classmethod) that you could add to the definition of `myclassmethod` to tell it - but it's a bit frustrating that it complains even if you try to add an explicit annotation. I guess it's preferring flagging up potential errors with missing `self` parameters over supporting more advanced use-cases?

**Comment by erictraut** [comment]
One question that you'd need to answer is: how would one correctly annotate the return type for `mystaticmethod` such that a type checker would know to treat it as a static method? If you use `Callable[P, R]`, it will be interpreted as an instance variable. There's currently no way in the type system to specify that you want to treat a callable as a static method — one that does not require any binding when referenced from a class or object.

Another aspect has to do with `ParamSpec` and what happens when a method is "captured" by a `ParamSpec`. Is the `self` or `cls` parameter also captured in the signature? Is the fact that the method is a static, class, or instance method captured? If the `ParamSpec` is later used to parameterize some class, is this method information retained by the specialized class?

There are probably other issues that I'm not able to think of off the top of my head. The reason this has gone unspecified (and unaddressed) for so long is that it's a large can of worms. It involves many issues, and all of them need to be considered.

**Comment by QuantumTim** [comment]
I'm not quite sure I follow. This was the implementation I knocked together to test it:

```python
_P = ParamSpec('_P')
_S_contra = TypeVar('_S_contra', contravariant=True)
_T_co = TypeVar('_T_co', covariant=True)

class bound_method(Generic[_S_contra, _P, _T_co]):
    def __init__(self, instance: _S_contra, fn: Callable[Concatenate[_S_contra, _P], _T_co]) -> None:
        self._instance = instance
        self._fn = fn
    def __call__(self, /, *args: _P.args, **kwargs: _P.kwargs) -> _T_co:
        self._instance: _S_contra
        self._fn: Callable[Concatenate[_S_contra, _P], _T_co]
        return self._fn(self._instance, *args, **kwargs)

class mystaticmethod(Generic[_P, _T_co]):
    def __init__(self, fn: Callable[_P, _T_co]) -> None:
        self._fn = fn
    def __get__(self, instance: Any, owner: type[Any]|None = None) -> Callable[_P, _T_co]:
        self._fn: Callable[_P, _T_co]
        return self._fn

class myclassmethod(Generic[_S_contra, _P, _T_co]):
    def __init__(self, fn: Callable[Concatenate[type[_S_contra], _P], _T_co]) -> None:
        self._fn = fn
    def __get__(self, instance: _S_contra|None, owner: type[_S_contra]|None = None) -> bound_method[type[_S_contra], _P, _T_co]:
        self._fn: Callable[Concatenate[type[_S_contra], _P], _T_co]
        if owner is None:
            assert instance is not None
            owner = type(instance)
        return bound_method(owner, self._fn)
```

If you use this with the definition of `class A` above, then everything "just works" (apart from the fact that pyright complains that my functions aren't using `self` correctly because it doesn't know any better).

> how would one correctly annotate the return type for mystaticmethod such that a type checker would know to treat it as a static method

The definition of the `__get__` descriptor solves this, no?

> Another aspect has to do with ParamSpec and what happens when a method is "captured" by a ParamSpec. Is the self or cls parameter also captured in the signature? Is the fact that the method is a static, class, or instance method captured? If the ParamSpec is later used to parameterize some class, is this method information retained by the specialized class?

With the above implementation you get the following, which seems correct/sensible to me:

```python
def g[**P, R](f: Callable[P, R]) -> Callable[P, R]:
    return f

reveal_type(g(A.clsmeth))    # Type of "g(A.clsmeth)" is "(a: int, *, b: str) -> list[bool]"
reveal_type(g(a.clsmeth))    # Type of "g(a.clsmeth)" is "(a: int, *, b: str) -> list[bool]"
reveal_type(g(A.staticmeth)) # Type of "g(A.staticmeth)" is "(a: int, *, b: str) -> set[bool]"
reveal_type(g(a.staticmeth)) # Type of "g(a.staticmeth)" is "(a: int, *, b: str) -> set[bool]"
```
